File: Programmers/[42894] 블록 게임/Johoseong/42894.py
def is_bomb(lst):
    row = set()
    col = set()
    for r, c in lst:
        if r not in row:
            row.add(r)
        if c not in col:
            col.add(c)
    answer = []
    for r in row:
        for c in col:
            if (r,c) not in lst:
                answer.append((r,c))
    return answer

def can_del(lst,black):
    for block in lst:
        if block not in black:
            return False
    return True        

def solution(board):
    N = len(board)
    answer = 0
    updated = True
    while updated:
        block_info = {}
        black = set()
        check = [False for _ in range(N)]
        for i in range(N):
            for j in range(N):
                typed = board[i][j]
                if typed != 0:
                    if typed in block_info:
                        block_info[typed][0].append((i,j))
                    else:
                        block_info[typed] = [[(i,j)],list()]
                    check[j] = True
                else:
                    if not check[j]:
                        black.add((i,j))
        for typed in block_info:
            block_info[typed][1] = is_bomb(block_info[typed][0])

        updated = False
        for typed in block_info:
            if can_del(block_info[typed][1],black):
                # 삭제 가능
                answer +=1
                updated = True
                for i,j in (block_info[typed][0] + block_info[typed][1]):
                    board[i][j] = 0
    return answer

File: Programmers/[42894] 블록 게임/KingDonggyu/42894.js
const blockPattern = [
  [[1, 0], [1, 1], [1, 2]],
  [[1, 0], [2, 0], [2, -1]],
  [[1, 0], [2, 0], [2, 1]],
  [[1, 0], [1, -1], [1, -2]],
  [[1, 0], [1, -1], [1, 1]],
];

const blockEmpty = [
  [[0, 1], [0, 2]],
  [[0, -1], [1, -1]],
  [[0, 1], [1, 1]],
  [[0, -1], [0, -2]],
  [[0, -1], [0, 1]]
];

function checkBlockAbove(board, x, y, type) {
  for (let i = 0; i < blockEmpty[type].length; i++) {
    const [dx, dy] = blockEmpty[type][i];
    const col = y + dy;
    for (let row = x + dx; row >= 0; row--) {
      if (board[row][col]) {
        return false;
      }
    }
  }

  return true;
}

function removeBlocks(board, x, y) {
  const number = board[x][y];
  
  for (let i = 0; i < 5; i++) {
    const matched = [[x, y]];

    for (const [dx, dy] of blockPattern[i]) {
      const nextX = x + dx;
      const nextY = y + dy;

      if (
        nextX < 0 || nextX >= board.length || 
        nextY < 0 || nextY >= board.length ||
        board[nextX][nextY] !== number
      ) {
        break;
      }

      matched.push([nextX, nextY]);
    }

    // 패턴 찾음
    if (matched.length === 4) {
      // 패턴 블록의 위에 블록이 있는지 확인
      if (!checkBlockAbove(board, x, y, i)) {
        break;
      }

      matched.forEach(([x, y]) => {
        board[x][y] = 0;
      })

      return 1;
    }
  }

  return 0;
}

function solution(board) {
  let answer = 0;

  board.forEach((row, x) => {
    row.forEach((block, y) => {
      if (block) {
        answer += removeBlocks(board, x, y);
      }
    })
    row.forEach((block, y) => {
      if (block) {
        answer += removeBlocks(board, x, y);
      }
    })
  });

  return answer;
}